### PR TITLE
feat(python): better async_collect

### PR DESCRIPTION
**Pull request by Object905**
In #10616 @jakob-keller proposed to simplify this to `await ldf.collect_async()` for asyncio. Not it's possible.

Api now is way nicer and defaults to asyncio, since its most common usage. Don't need to pass Queue instances.
Also since they had to be thread safe - default queues from asyncio and gevent were not compatible and external dependency janus was required to safely use them.

asyncio implementation turned out fairly simple 
But with gevent  I was not able to to do this with [ILoop.run_callback_threadsafe](https://www.gevent.org/api/gevent.hub.html#gevent._interfaces.ILoop.run_callback_threadsafe) for some reason due to `LoopExit: This operation would block forever`, while they say it's a [asyncio.loop.call_soon_threadsafe](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_soon_threadsafe) alternative. 
Maybe there are some checks that are not compatible with how objects are called from out of thread rust side...

Managed to use lower level api, but with nuance that results into AsyncResult are only being set when context switch occurs. (get(block=True), get(block=False, timeout=...) works, but get(block=False, timeout=None) does not),
worked around making result a property to check for this case.

### CODE REVIEW

**Review by stinodego — COMMENTED** [review]
I don't really know enough about async to judge this PR on its merits, but form-wise it looks great with updated docs and everything.

Looks like the method doesn't have a Parameters section in the docstring though. That could be useful to document this new parameter properly.

**Comment by ritchie46** [comment]
Thank you @Object905 can you do a rebase?

**Comment by Object905** [comment]
Sure, will do, a bit later. 

Also want to write some test, because there were some corner cases that i've tested manually, especially in gevent, where it has different loop implementations and defaults to different one on windows. Seems like it wont be a problem given how public api documented, but who knows.

**Comment by Object905** [comment]
Ready for review

File: py-polars/requirements-dev.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ connectorx
 cloudpickle
 fsspec
 s3fs[boto3]
+gevent
```
Comment (stinodego) [inline py-polars/requirements-dev.txt:24]: We have a `dependencies` module for handling optional dependencies.

We should probably have an `async` option then in `pyproject.toml` then with `gevent` as well.

Comment (Object905) [reply py-polars/requirements-dev.txt:24]: I'll add `gevent`, but don's see any use in `async`, since it only uses standard library.

Comment (Object905) [reply py-polars/requirements-dev.txt:24]: Added gevent to dependencies.py, made better ImportError.

File: py-polars/pyproject.toml [hunk 2]
```diff
@@ -51,9 +51,9 @@ pydantic = ["pydantic >= 2.0.0"]
 sqlalchemy = ["sqlalchemy", "pandas"]
 xlsxwriter = ["xlsxwriter"]
 adbc = ["adbc_driver_sqlite"]
-cloudpickle = ["cloudpickle"]
```
Comment (jakob-keller) [inline py-polars/pyproject.toml:54]: The `cloudpickle` optional dependency was replaced. I believe that was unintended and should be reverted.

Comment (Object905) [reply py-polars/pyproject.toml:54]: Yep, fixed

File: py-polars/polars/dependencies.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
 _PYARROW_AVAILABLE = True
 _PYDANTIC_AVAILABLE = True
 _ZONEINFO_AVAILABLE = True
+_GEVENT_AVAILABLE = True
```
Comment (jakob-keller) [inline py-polars/polars/dependencies.py:20]: Might want to move this up to line 14 to keep globals sorted alphabetically.

File: py-polars/polars/dependencies.py [hunk 4]
```diff
@@ -232,6 +235,7 @@ def _check_for_pydantic(obj: Any) -> bool:
     "pydantic",
     "pyarrow",
     "zoneinfo",
+    "gevent",
```
Comment (jakob-keller) [inline py-polars/polars/dependencies.py:238]: Might want to move this up to line 233 to keep list (section) sorted alphabetically.

File: py-polars/polars/dependencies.py [hunk 5]
```diff
@@ -246,4 +250,5 @@ def _check_for_pydantic(obj: Any) -> bool:
     "_PANDAS_AVAILABLE",
     "_PYARROW_AVAILABLE",
     "_ZONEINFO_AVAILABLE",
+    "_GEVENT_AVAILABLE",
```
Comment (jakob-keller) [inline py-polars/polars/dependencies.py:253]: Might want to move this up to line 248 to keep list (section) sorted alphabetically.

File: py-polars/polars/functions/lazy.py [hunk 6]
```diff
@@ -1664,33 +1699,46 @@ def collect_all_async(
...
-) -> _AsyncDataFrameResult[list[DataFrame]]:
+) -> Awaitable[list[DataFrame]] | _GeventDataFrameResult[list[DataFrame]]:
     """
     Collect multiple LazyFrames at the same time asynchronously in thread pool.
 
     Collects into a list of DataFrame, like :func:`polars.collect_all`
-    but instead of returning them directly its collected inside thread pool
-    and gets put into `queue` with `put_nowait` method,
-    while this method returns almost instantly.
+    but instead of returning them directly they are scheduled to be collected
```
Comment (jakob-keller) [inline py-polars/polars/functions/lazy.py:1707]: add comma: "directly, they are"

File: py-polars/polars/lazyframe/frame.py [hunk 7]
```diff
@@ -1707,33 +1740,44 @@ def collect_async(
...
         streaming: bool = False,
-    ) -> _AsyncDataFrameResult[DataFrame]:
+    ) -> Awaitable[DataFrame] | _GeventDataFrameResult[DataFrame]:
         """
         Collect DataFrame asynchronously in thread pool.
 
-        Collects into a DataFrame, like :func:`collect`
-        but instead of returning DataFrame directly its collected inside thread pool
-        and gets put into `queue` with `put_nowait` method,
+        Collects into a DataFrame, like :func:`collect` but instead of returning
```
Comment (jakob-keller) [inline py-polars/polars/lazyframe/frame.py:1747]: Maybe adjust punctuation: "Collects into a DataFrame (like :func:`collect`), but instead of returning dataframe directly, they are scheduled to be collected inside thread pool, ..."

File: py-polars/polars/utils/_async.py [hunk 8]
```diff
@@ -1,45 +1,92 @@
...
 
     def _callback_all(self, obj: list[PyDataFrame] | Exception) -> None:
         if not isinstance(obj, Exception):
             obj = [wrap_df(pydf) for pydf in obj]
-        self.queue.put_nowait(obj)  # type: ignore[arg-type]
+        self._value = obj
+        self._watcher.send()
+
+
+class _AioDataFrameResult(Generic[T], Awaitable[T]):
```
Comment (jakob-keller) [inline py-polars/polars/utils/_async.py:67]: By convention, I would flip both parent classes, i.e. `class _AioDataFrameResult(Awaitable[T], Generic[T]):`

File: py-polars/polars/utils/show_versions.py [hunk 9]
```diff
@@ -71,6 +71,7 @@ def _get_dependency_info() -> dict[str, str]:
         "sqlalchemy",
         "xlsx2csv",
         "xlsxwriter",
+        "gevent",
```
Comment (jakob-keller) [inline py-polars/polars/utils/show_versions.py:74]: Might want to move this up to line 66 to keep list sorted alphabetically.

File: py-polars/pyproject.toml [hunk 10]
```diff
@@ -86,6 +87,7 @@ module = [
   "xlsx2csv",
   "xlsxwriter.*",
   "zoneinfo",
+  "gevent",
```
Comment (jakob-keller) [inline py-polars/pyproject.toml:90]: Might want to move this up to line 81 to keep list sorted alphabetically.

File: py-polars/tests/unit/test_async.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    [
+        (_aio_collect_async, None),
+        (_aio_collect_all_async, None),
+        (partial(_aio_collect_async, True), pl.ColumnNotFoundError),
+        (partial(_aio_collect_all_async, True), pl.ColumnNotFoundError),
+    ],
+)
+
+
+def _aio_run(coroutine: Any, raises: Exception | None = None) -> None:
```
Comment (jakob-keller) [inline py-polars/tests/unit/test_async.py:62]: Have you looked into https://github.com/pytest-dev/pytest-asyncio for running asyncio tests? I find it very helpful to reduce boilerplate code.

**Review by jakob-keller — COMMENTED** [review]
Thanks for picking this up. I can't wait to give it a try!

**Comment by ritchie46** [comment]
@Object905 can you do another pass for @jakob-keller's review? Then we can work to merging this.

**Comment by Object905** [comment]
Sorry for the delay. Fixed everything regarding punctuation and alphabetic sorting.

Regarding [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio). I don't think its justified for such small test suite. Also because tests are parametrized for exception this wrapper does more than just wraps asyncio.run. 
Similar wrapper _gevent_run is needed for gevent.

**Review by jakob-keller — COMMENTED** [review]
LGTM, great job @Object905!

I also did a tiny bit of testing of last week's version and it passed my internal test suite. I am planning to use this in real life, once released.

**Comment by ritchie46** [comment]
Thanks for the PR @Object905  and thanks for the review @jakob-keller. Much appreciated as I am not really familiar with this.

**Comment by jakob-keller** [comment]
Time for a release? 🙂

**Comment by winstxnhdw** [comment]
No tests to check if it's non-blocking? On sufficiently complex queries, `collect_async` still blocks the main event loop for me.
